contracts/Compressor.sol

```diff
@@ -0,0 +1,250 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {ICompressor, OPERATION_BITMASK, LENGTH_BITS_OFFSET, MAX_ENUMERATION_INDEX_SIZE} from "./interfaces/ICompressor.sol";
+import {ISystemContract} from "./interfaces/ISystemContract.sol";
+import {Utils} from "./libraries/Utils.sol";
+import {UnsafeBytesCalldata} from "./libraries/UnsafeBytesCalldata.sol";
+import {EfficientCall} from "./libraries/EfficientCall.sol";
+import {L1_MESSENGER_CONTRACT, INITIAL_WRITE_STARTING_POSITION, COMPRESSED_INITIAL_WRITE_SIZE, STATE_DIFF_ENTRY_SIZE, STATE_DIFF_ENUM_INDEX_OFFSET, STATE_DIFF_FINAL_VALUE_OFFSET, STATE_DIFF_DERIVED_KEY_OFFSET, DERIVED_KEY_LENGTH, VALUE_LENGTH, ENUM_INDEX_LENGTH, KNOWN_CODE_STORAGE_CONTRACT} from "./Constants.sol";
+
+/**
+ * @author Matter Labs
+ * @custom:security-contact security@matterlabs.dev
+ * @notice Contract with code pertaining to compression for zkEVM; at the moment this is used for bytecode compression
+ * and state diff compression validation.
+ * @dev Every deployed bytecode/published state diffs in zkEVM should be publicly restorable from the L1 data availability.
+ * For this reason, the user may request the sequencer to publish the original bytecode and mark it as known.
+ * Or the user may compress the bytecode and publish it instead (fewer data onchain!). At the end of every L1 Batch
+ * we publish pubdata, part of which contains the state diffs that occurred within the batch.
+ */
+contract Compressor is ICompressor, ISystemContract {
+    using UnsafeBytesCalldata for bytes;
+
+    /// @notice Verify the compressed bytecode and publish it on the L1.
+    /// @param _bytecode The original bytecode to be verified against.
+    /// @param _rawCompressedData The compressed bytecode in a format of:
+    ///    - 2 bytes: the length of the dictionary
+    ///    - N bytes: the dictionary
+    ///    - M bytes: the encoded data
+    /// @dev The dictionary is a sequence of 8-byte chunks, each of them has the associated index.
+    /// @dev The encoded data is a sequence of 2-byte chunks, each of them is an index of the dictionary.
+    /// @dev The compression algorithm works as follows:
+    ///     1. The original bytecode is split into 8-byte chunks.
+    ///     Since the bytecode size is always a multiple of 32, this is always possible.
+    ///     2. For each 8-byte chunk in the original bytecode:
+    ///         * If the chunk is not already in the dictionary, it is added to the dictionary array.
+    ///         * If the dictionary becomes overcrowded (2^16 + 1 elements), the compression process will fail.
+    ///         * The 2-byte index of the chunk in the dictionary is added to the encoded data.
+    /// @dev Currently, the method may be called only from the bootloader because the server is not ready to publish bytecodes
+    /// in internal transactions. However, in the future, we will allow everyone to publish compressed bytecodes.
+    function publishCompressedBytecode(
```

Comment:
I don't think we added `require(_bytecode.length >= _rawCompressedData.length);`

Reply:
We'll add it in the next release